### DISCUSSION TITLE
usbgpib_conf.py a quick hack using python-usbtmc

### DISCUSSION THREAD

**Discussion by xyphro** [xyphro/UsbGpib#55] [Ideas]
Here a nice script shared by user: [jimthouston](https://github.com/jimthouston), that I'd like to store here in the discussion area.

I have been playing with UsbGpib for a couple months.  I'm a Linux user and have been using pyvisa with the pyvisa-py back end.  This doesn't support the control_in method.

I found that python-usbtmc works nicely for sending the config options.

$ cat usbgpib_conf.py
```
import sys
import usbtmc

inst = usbtmc.Instrument(0x3eb, 0x2065)
if len(sys.argv) < 2:
    print('Usage: usbgpib_conf command')
    for cmd in [ '!autoid?', '!term?', '!string?', '!ver?' ]:
        inst.pulse()
        print(cmd, inst.ask(cmd))
    sys.exit(0)

cmd = '!' + ' '.join(sys.argv[1:])
print(cmd)
inst.pulse()
if cmd[-1] == '?':
    print(inst.ask(cmd))
else:
    inst.write(cmd)
```

$ python3 usbgpib_conf.py
Usage: usbgpib_conf command
!autoid? off
!term?  eoi
!string? normal
!ver? V1.6

$ python3 usbgpib_conf.py autoid on

**Comment by rjordans** [comment]
Thanks, this helped a lot!

For those working under Linux it may be good to note that (by default) only root has access to the USB devices.  I've added the following to my udev rules to allow group access for those in the plugdev group.

File ```/etc/udev/rules.d/9-usbtmc.rules```:
```
SUBSYSTEM=="usb", ATTR{idVendor}=="03eb", ATTR{idProduct}=="2065", MODE="0660", GROUP="plugdev"
KERNEL=="usbtmc*", MODE="0660", GROUP="plugdev"
```

**Reply by jimthouston** [reply]
I have been playing with nicegui and have a first cut a gui for the usbgpib configuration
settings.  Nicegui lets you implement a gui in python using the browser to handle
the display.  It is also supposed to work with a native window but I have not figured out the prerequisites.  See the nicegui documentation here:
https://nicegui.io/
https://github.com/zauberzeug/nicegui

```
$ cat usbgpib_gui.py
# nicegui interface to configure the xyphro UsbGpib settings
# This needs work to support more than one UsbGpib interface.
# Jim Houston

from nicegui import ui
import usbtmc

cmd_list = [
    {'desc':'String length:', 'cmd':'!string',
    'values':['normal', 'short'] },
    {'desc':'Auto ID:', 'cmd':'!autoid',
    'values': ['off', 'on', 'slow', 'slower', 'slowest'] },
    {'desc':'Read terminator:', 'cmd':'!term',
    'values':['lf', 'cr', 'eoi', 'store'] }
]

def get_value(c):
    inst = usbtmc.Instrument(0x3eb, 0x2065)
    inst.pulse()
    v = inst.ask(c + '?')
    inst.close()
    return(v)

def set_value(cmd, v):
    inst = usbtmc.Instrument(0x3eb, 0x2065)
    inst.pulse()
    inst.write(f'{cmd} {v}')
    inst.close()
    usbgpib_conf.refresh()

@ui.refreshable
def usbgpib_conf() -> None:
    ui.label('Usbgpib configuration GUI')
    with ui.grid(columns=2):
        ui.label('Firmware version:')
        ui.label(get_value('!ver?'))
        for c in cmd_list:
            v = get_value(c['cmd'])
            ui.label(c['desc'])
            sel = ui.select(options=c['values'], value=v,
                on_change= lambda e: set_value(c['cmd'], e.value))
            sel.c = c

usbgpib_conf()

ui.run()
```

**Reply by jimthouston** [reply]
I have updates for both of my scripts.  The original versions only worked with a single UsbGpib board.


I tested on Ubuntu 24.04 installing the prerequisites in a virtual environment.  In theory the usbtmc and
nicegui libraries are portable.

Here is the minimum setup to use the GUI with nicegui using the browser:
```
python3 -m venv venv
source venv/bin/activate
pip3 install nicegui
pip3 install starlette==0.41.3
pip3 install python-usbtmc pyusb
```
The install of starlette==0.41.3 is a workaround for a recent problem and would not normally be needed.
Using the python-usbtmc library needs the udev setup which rjordans describes above.

Save the following as usbgpib_gui.py:

```
# nicegui interface to configure the xyphro UsbGpib settings
# Jim Houston

import multiprocessing
multiprocessing.set_start_method("spawn", force=True)
from multiprocessing import Process, Lock

from nicegui import ui, app, run
import usbtmc
import usb.core
import usb.util

cmd_list = [
    {'desc':'Firmware version:', 'cmd':'!ver' },
    {'desc':'String length:', 'cmd':'!string',
    'values':['normal', 'short'] },
    {'desc':'Auto ID:', 'cmd':'!autoid',
    'values': ['off', 'on', 'slow', 'slower', 'slowest'] },
    {'desc':'Read terminator:', 'cmd':'!term',
    'values':['lf', 'cr', 'eoi', 'store'] }
]

dev_sel = None
sel = {}
lab = {}
lock = Lock()

def open_inst(sel):
    # print(f'open_inst(sel={sel})')
    if not sel or not sel.value:
        print(f'No serial number selected')
        return None
    inst = usbtmc.Instrument(0x3eb, 0x2065, sel.value)
    if not inst:
        print(f'Failed to open instrument {sel.value}')
    return inst

def set_value(cmd, v):
    # print(f'set_value({cmd}, {v})')
    lock.acquire()
    try:
        inst = open_inst(dev_sel)
        if not inst:
            return
        inst.pulse()
        discard = inst.ask(f'{cmd} {v}')
        # read back the value.  When we select store I want to go
        # back to showing the term value.
        inst.pulse()
        v = inst.ask(cmd + '?')
        inst.close()
    finally:
        lock.release()
    return v

async def update_value(e):
    # print(f'update_value({e})')
    sel = e.sender
    v = e.value
    cmd = sel.cmd
    # on_change is called when we set the initial value sigh.
    # Don't update the hardware if nothing has changed.
    # print(f'cmd = {cmd} v = {v} sel.value = {sel.value}')
    # print(f'sel.last_val = {sel.last_val}')
    if  sel.last_val == v:
        return
    v = await run.io_bound(set_value, cmd, v)
    if not v:
        return
    if sel.value != v:
        sel.value = v
        sel.last_val = v

def do_exit():
    ui.run_javascript('window.close()')
    app.shutdown()

def get_values():
    # print(f'get_values')
    lock.acquire()
    try:
        inst = open_inst(dev_sel)
        if not inst:
            return
        v = {}
        for c in cmd_list:
            cmd = c['cmd']
            inst.pulse()
            v[cmd] = inst.ask(cmd + '?')
        inst.close()
    finally:
        lock.release()
    return(v)

async def async_get_values():
    v = await run.io_bound(get_values)
    # print(f'v = {v}')
    if v:
        for cmd in v.keys():
            if cmd in sel.keys():
                sel[cmd].last_val = v[cmd]
                sel[cmd].value = v[cmd]
            if cmd in lab.keys():
                lab[cmd].text = v[cmd]

# I poll to detect USB insert/removal.  There usbtmc python
# library doesn't detect that the usbgpib has been removed.

def list_devs():
#    devices = usbtmc.list_devices()
#    l = []
#    for d in devices:
#        if d.idVendor == 0x3eb and d.idProduct == 0x2065:
#            try:
#                l.append(d.serial_number)
#            except ValueError:
#                print(f'list_devs: got ValueError exception')
# find our device
    lock.acquire()
    try:
        l = []
        for d  in usb.core.find(idVendor=0x03eb, idProduct=0x2065, find_all=True):
            #langids = d.langids
            #print(langids)
            #serial = usb.util.get_string(d, d.iSerialNumber, langids[0])
            serial = usb.util.get_string(d, d.iSerialNumber, 1033)
            l.append(serial)
    finally:
        lock.release()

    return l

async def update_devices(sel):
    #l = await run.cpu_bound(list_devs)
    l = list_devs()
    if set(l) == set(sel.options):
        return
    v = sel.value
    # pick the first serial number if we don't have a useable value
    # if l and not sel.value in l:
    if l != [] and (sel.value == None or (not sel.value in l)):
        sel.set_options(l, value=l[0])
    else:
        sel.set_options(l)

@ui.page("/")
async def usbgpib_conf() -> None:
    global dev_sel
    ui.label('Usbgpib configuration GUI')
    ui.button('Exit', on_click = lambda: do_exit())

    c1 = ui.card()
    c2 = ui.card()
    
    with c2:
        with ui.grid(columns=2):
            for c in cmd_list:
                cmd = c['cmd']
                ui.label(c['desc'])
                if 'values' in c.keys():
                    sel[cmd] = ui.select(options=c['values'],
                        on_change= update_value)
                    sel[cmd].cmd = cmd
                else:
                    lab[cmd] = ui.label()
    
    with c1:
        dev_sel = ui.select([], label='Select device serial number',
            on_change= lambda: async_get_values())
    
    ui.timer(1.0, lambda: update_devices(dev_sel))



ui.run(reload=False)
#ui.run(native=True, window_size=(400, 500), fullscreen=False,reload=False)
```
This script will also work if the native=True option is used on the ui.run() call.  This requires
additional setup as follows:

```
pip3 install nicegui pywebview qtpy pyqt5 PyQtWebEngine
sudo apt install python3-gi python3-gi-cairo gir1.2-gtk-3.0 gir1.2-webkit2-4.1
sudo apt-get install ^libxcb.*-dev libx11-xcb-dev libglu1-mesa-dev libxrender-dev libxi-dev libxkbcommon-dev libxkbcommon-x11-dev
```

Here are screenshots:
![p1](https://github.com/user-attachments/assets/91418e6d-6294-4f5d-9f58-0e3493fc504b)
![p2](https://github.com/user-attachments/assets/e0af1bb0-246f-4f8e-be69-0239df94d6ce)

Here is the updated usbgpib_conf.py:
```import sys
import usbtmc
import sys 
import getopt 
  
def display_conf(serial):
    print(f'UsbGpib config for {serial}')
    inst = usbtmc.Instrument(0x3eb, 0x2065, serial)
    for cmd in [ '!autoid?', '!term?', '!string?', '!ver?' ]:
        inst.pulse()
        print(cmd, inst.ask(cmd))
    inst.close()

def send_cmd(serial, cmd):
    inst = usbtmc.Instrument(0x3eb, 0x2065, serial)
    inst.pulse()
    if cmd[-1] == '?':
        print(inst.ask(cmd))
    else:
        inst.write(cmd)
    inst.close()


def display_all():
    devices = usbtmc.list_devices()
    for d in devices:
        if d.idVendor == 0x3eb and d.idProduct == 0x2065:
            display_conf(d.serial_number)

def usbgpib_conf():
    serial = None
    argv = sys.argv[1:] 
  
    try: 
        opts, args = getopt.getopt(argv, "s:") 
      
    except: 
        print("Error") 
  
    for opt, arg in opts: 
        if opt in ['-s']: 
            serial = arg 

    if len(args) == 0:
        if serial:
            display_conf(serial)
        else:
            display_all()
        return

    cmd = '!' + ' '.join(args)
    print(cmd)
    send_cmd(serial, cmd)
      
usbgpib_conf()     
```
Here is an example running the script:
```
(venv) jhouston@T5610:~/UsbGpib_new$ python3 usbgpib_conf.py
UsbGpib config for GPIB_22_342313232313
!autoid? off
!term? cr
!string? normal
!ver? V1.6
UsbGpib config for HEWLETT-PACKARD_6642A_0_fA.01.08sA.01.03pA.01.06
!autoid? on
!term? eoi
!string? normal
!ver? V1.8
(venv) jhouston@T5610:~/UsbGpib_new$ python3 usbgpib_conf.py -s GPIB_22_342313232313 term eol
!term eol
```

Lessons learned:
It is easy to forget that the usbtmc library is implemented in user code.  Running multiple copies at the same time is a bad idea.  I was surprised that the nicegui framework was happily getting the list of serial numbers and at the same time reading the values of the parameters.  It took me a while to figure this out from the error "ValueError: The device has no langid" .  Adding the locking to the usbgpib_gui.py fixed the immediate problem.  It doesn't solve the larger problem which will occur if multiple processes are accessing the same devices at the same time.

Kai I hope you will add these scripts to the source.  I'm sorry they are arriving after you wrote your own GUI.

Jim

**Reply by xyphro** [reply]
Hi Jim,

Excellent. Don't worry, I don't see this as double work cause some users not using python simply prefer an .exe, eg. Labview guys often have now Python installation :-)

Can put this in the repository. Any preference for the location?

Best regards,

Kai

**Reply by jimthouston** [reply]
Hi Kai,

Maybe a 'scripts' directory either under SW or at the top level?  The important part is to have them under source control so people find the current version with out having to scan the discussion.

Jim

**Reply by dazz100** [reply]
I suggest a folder that describes what it does (USB-GPIB configuration app), not what it is (a script).

**Reply by jimthouston** [reply]
Hi Kai,

I figured out why my polling for UsbGpib devices gets errors when I read the parameters.  The UsbGpib firmware is just a polling loop and it does a 1/4 second delay loop to implement the pulse indicator command.  While it delays it doesn't answer USB requests for things like the langid value resulting in the ValueError exception from the pyusb library.

I'm not sure it hurts anything.  Maybe the pulse command could simply turn on the LED and it could be turned off after a few thousand times through the main loop.

Jim

**Comment by xyphro** [comment]
Please check it here:
https://github.com/xyphro/UsbGpib?tab=readme-ov-file#python-gui

Let me know if you agree to the text and folder naming.

@jimthouston: I think I should replace the blocking blink delays and make the blink asynchronous then in the next release. Have put it on my todo list. That's a simple one.

**Reply by jimthouston** [reply]
Hi Kai,

It would be nice to mention Nicegui in the toplevel readme as a shout out.

I have been trying not to over promise on the portability of the python GUI.  I have only tested on Linux.  I would like to know if folks get it to work on other platforms.

I assume that the new directory will have both the usbgpib_gui.py and the usbgpib_conf.py.  It would make sense to have a readme which would be the cleaned up text from the discussion to cover the dependencies and running under linux.

Jim

**Reply by xyphro** [reply]
Hi Jim,

Can you draft such a readme? I cannot test the gui myself right now not depend too much time on following the thread.

I can announce it as linux only solution then.

BR,

Kai

**Reply by jimthouston** [reply]
Hi Kai,

I will do a Readme.md file.  I assume that the scripts and two screenshots could go in the same directory.

Jim

**Comment by xyphro** [comment]
Thanks, it does not need to be in markdown format, just a plain text is ok
- i can format it. For the picture an [insert picture xyz here] is good
enough.

jimthouston ***@***.***> schrieb am Di., 28. Jan. 2025, 18:53:

**Reply by jimthouston** [reply]
Hi Kai,

I'm attaching a readme.md.gz, usbgpib_conf.py.gz, p1.png, p2.png.
I hope gzip works for you.  The pictures and script are the same as in the discussion.

Jim

![p1](https://github.com/user-attachments/assets/eda57b40-fd18-4590-b7f2-4c2248d87075)
![p2](https://github.com/user-attachments/assets/779eeae7-c6c9-46d0-a36d-6e6cdca983b4)
[readme.md.gz](https://github.com/user-attachments/files/18579092/readme.md.gz)
[usbgpib_conf.py.gz](https://github.com/user-attachments/files/18579147/usbgpib_conf.py.gz)

**Comment by xyphro** [comment]
Hi Jim, thanks a lot. Gzip is surely no issue for me. Will update tomorrow
evening (= in abt 24h).
Best regards, Kai

jimthouston ***@***.***> schrieb am Di., 28. Jan. 2025, 21:55:

**Comment by xyphro** [comment]
Have just uploaded the files. Thanks for contributing!

**Comment by xyphro** [comment]
Jim - 

Next FW release 2.0 will support 2 additional options:
autoid slower
autoid slowest
Those were mentioned before, but did not really work as I noticed recently.

Furthermore the pulse indicator request call will work asynchronously. I could reproduce in some corner cases issues and could remove them by removing the blocking 250ms delay.
The USBTMC spec does not enforce timings, so visa/tmc implementations make assumptions on necessary timeout values. 250ms delay was just on the edge - when calling the indicator pulse request  function 2 times after each other with minimal delay, issues popped up for the response on the 2nd call (noticed by the fact that 0 bytes were returns and a timeout error gets reported).

**Reply by xyphro** [reply]
...and it is released.

I also included a better (additional) usbtmc compliant method for setting read termination. Somehow missed this when studying the standard earlier.
